### PR TITLE
feat!: redesigned Snakemake API. It now uses a modern, dataclass based approach

### PR DESCRIPTION
**Pull request by johanneskoester**
### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [ ] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### CODE REVIEW

File: snakemake/cli.py [hunk 1]
```diff
@@ -2480,3 +2481,120 @@ def print_candidates(candidates):
...
+                "--force",
+                "--keep-target-files",
+                "--keep-remote",
+                "--max-inventory-time 0",
+                "--nocolor",
+                "--notemp",
+                "--no-hooks",
+                "--nolock",
+                "--ignore-incomplete",
+                w2a("execution_settings.keep_incomplete")
```
Comment (Hocnonsense) [inline snakemake/cli.py:2561]: It seems that a comma is lost here

**Comment by kyleabeauchamp** [comment]
I *think* the recommendation for migration involves the `local` flag but it might be useful to add to the 7->8 migration guide.

**Comment by kyleabeauchamp** [comment]
Actually it's a find and replace of `srcdir` with `workflow.source_path`.  Probably good adding to the migration guide.

**Comment by claczny** [comment]
> Actually it's a find and replace of `srcdir` with `workflow.source_path`. Probably good adding to the migration guide.

Either I am doing something wrong or this is not a drop-in replacement.
I worked around it via `workflow.basedir`, e.g., `ENVS = Path(workflow.basedir) / "envs".

**Pull request merged**